cpp-client/build.gradle

```diff
@@ -108,19 +110,36 @@ def testCppClient = Docker.registerDockerTask(project, 'testCppClient') {
                       cd ..; \\
                       rm -fr ${PREFIX}/src/deephaven/build
                    ''')
+        // Note environment variables defined here are inherited by other images
+        // using this image as a base.
+        environmentVariable 'DH_PREFIX', prefix
+        environmentVariable 'LD_LIBRARY_PATH', "${prefix}/lib"
+    }
+    parentContainers = [ Docker.registryTask(project, 'r-client-base') ]
+}
+
+def testCppClient = Docker.registerDockerTask(project, 'testCppClient') {
+    // Only tested on x86-64, and we only build dependencies for x86-64
+    platform = 'linux/amd64'
+    copyIn {}
+    copyOut {
+        into layout.buildDirectory.dir('test-results')
+    }
+    dockerfile {
+        from('deephaven/cpp-client:local-build')
         //
         // Setup for test run.
         //
         environmentVariable 'DH_HOST', deephavenDocker.containerName.get()
         environmentVariable 'DH_PORT', '10000'
-        environmentVariable 'DH_PREFIX', prefix
-        environmentVariable 'LD_LIBRARY_PATH', "${prefix}/lib"
     }
     containerDependencies.dependsOn = [deephavenDocker.healthyTask]
     containerDependencies.finalizedBy = deephavenDocker.endTask
     network = deephavenDocker.networkName.get()
-    parentContainers = [ Docker.registryTask(project, 'cpp-client-base') ]
+    parentContainers = [ project.tasks.getByName('cppClient') ]
     entrypoint = ["${prefix}/bin/dhcpp/cpp-tests-to-junit.sh", '/out/cpp-test.xml', '/out/cpp-test.log']
 }
+tasks.getByName('testCppClient').dependsOn project.tasks.getByName('cppClient')
```

Comment:
when you do this, you forcibly initialize both of these tasks regardless of whether either of them will actually be running.

Instead consider updating the closure sent to `Docker.registerDockerTask(project, 'testCppClient') {}` (circa line 123)  to additionally do:
```
dependsOn "$project.path:cppClient"
```
or
```
dependsOn buildCppClientImage
```
(the latter references the TaskProvider return value of `Docker.registerDockerTask`

Reply:
```
cfs@caicai 23:28:59 ~/dh/oss1/deephaven-core
$ git diff
diff --git a/cpp-client/build.gradle b/cpp-client/build.gradle
index 358337b1f8..d0944760a5 100644
--- a/cpp-client/build.gradle
+++ b/cpp-client/build.gradle
@@ -119,6 +119,7 @@ def buildCppClientImage = Docker.registerDockerTask(project, 'cppClient') {
 }
 
 def testCppClient = Docker.registerDockerTask(project, 'testCppClient') {
+    dependsOn "$project.path:cppClient"
     // Only tested on x86-64, and we only build dependencies for x86-64
     platform = 'linux/amd64'
     copyIn {}
@@ -139,7 +140,6 @@ def testCppClient = Docker.registerDockerTask(project, 'testCppClient') {
     parentContainers = [ project.tasks.getByName('cppClient') ]
     entrypoint = ["${prefix}/bin/dhcpp/cpp-tests-to-junit.sh", '/out/cpp-test.xml', '/out/cpp-test.log']
 }
-tasks.getByName('testCppClient').dependsOn project.tasks.getByName('cppClient')
 
 deephavenDocker.shouldLogIfTaskFails testCppClient
 tasks.check.dependsOn(testCppClient)

cfs@caicai 23:29:02 ~/dh/oss1/deephaven-core
$ ./gradlew :cpp-client:testCppClient

FAILURE: Build failed with an exception.

* Where:
Build file '/home/cfs/dh/oss1/deephaven-core/cpp-client/build.gradle' line: 122

* What went wrong:
A problem occurred evaluating project ':cpp-client'.
> Could not find method dependsOn() for arguments [:cpp-client:cppClient] on object of type Docker$DockerTaskConfig.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.6.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 312ms
10 actionable tasks: 10 up-to-date
```

And

```
cfs@caicai 23:29:55 ~/dh/oss1/deephaven-core
$ git diff
diff --git a/cpp-client/build.gradle b/cpp-client/build.gradle
index 358337b1f8..e621e47126 100644
--- a/cpp-client/build.gradle
+++ b/cpp-client/build.gradle
@@ -119,6 +119,7 @@ def buildCppClientImage = Docker.registerDockerTask(project, 'cppClient') {
 }
 
 def testCppClient = Docker.registerDockerTask(project, 'testCppClient') {
+    dependsOn buildCppClientImage
     // Only tested on x86-64, and we only build dependencies for x86-64
     platform = 'linux/amd64'
     copyIn {}
@@ -139,7 +140,6 @@ def testCppClient = Docker.registerDockerTask(project, 'testCppClient') {
     parentContainers = [ project.tasks.getByName('cppClient') ]
     entrypoint = ["${prefix}/bin/dhcpp/cpp-tests-to-junit.sh", '/out/cpp-test.xml', '/out/cpp-test.log']
 }
-tasks.getByName('testCppClient').dependsOn project.tasks.getByName('cppClient')
 
 deephavenDocker.shouldLogIfTaskFails testCppClient
 tasks.check.dependsOn(testCppClient)

cfs@caicai 23:29:57 ~/dh/oss1/deephaven-core
$ ./gradlew :cpp-client:testCppClient

FAILURE: Build failed with an exception.

* Where:
Build file '/home/cfs/dh/oss1/deephaven-core/cpp-client/build.gradle' line: 122

* What went wrong:
A problem occurred evaluating project ':cpp-client'.
> Could not find method dependsOn() for arguments [provider(task 'cppClient', class org.gradle.api.DefaultTask)] on object of type Docker$DockerTaskConfig.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.6.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 301ms
10 actionable tasks: 10 up-to-date
```

Which essentially are the same error.

Reply:
I think this is what he's suggesting:
```patch

diff --git a/cpp-client/build.gradle b/cpp-client/build.gradle
index 358337b1f8..332726996c 100644
--- a/cpp-client/build.gradle
+++ b/cpp-client/build.gradle
@@ -139,7 +139,9 @@ def testCppClient = Docker.registerDockerTask(project, 'testCppClient') {
     parentContainers = [ project.tasks.getByName('cppClient') ]
     entrypoint = ["${prefix}/bin/dhcpp/cpp-tests-to-junit.sh", '/out/cpp-test.xml', '/out/cpp-test.log']
 }
-tasks.getByName('testCppClient').dependsOn project.tasks.getByName('cppClient')
+testCppClient.configure {
+    dependsOn buildCppClientImage
+}
 
 deephavenDocker.shouldLogIfTaskFails testCppClient
 tasks.check.dependsOn(testCppClient)
```

Reply:
colin's comment is the shortest form to make it do what I was suggesting.  I didn't understand the purpose of the closure passed to Docker.registerDockerTask properly, I thought it was applied to a task and executed, which would provide `dependsOn`, but colin's answer sidesteps that and gets to the meat of the issue w/ the least complexity / effort.

Reply:
Done, it worked, thank you guys.

Reply:
(We would love to have it be a proper task configuring closure, but as far as I can tell it is impossible to define a task as "run these tasks", there's no such thing as task composition to produce a task. As we need to provide different configuration details to different constituent tasks, another option could be to define an extension... but as in projects like this one, we need to define at least two instances of "do stuff with docker, using inputs and outputs, etc.." as an extension... and i'm not aware of a nice way to apply the same extension more than once. So, we return the "most-downstream" task from the function, and as needed have other tasks depend on it.)

Yep  I'm sorry, what I gave you is equivalent to what you had before, but also wrong - the testCppClient task is the very last step in "run the test in docker and then finally copy the results out to gradle", which means that this dependency will not make sense, Gradle would be within its rights to not build the cpp client image until just before R test output is attempted to be copied from docker.

However, if I am reading this all right, it is also unnecessary - this line already covers the requirement you're after:
```groovy
    parentContainers = [ project.tasks.getByName('cppClient') ]
```
This ensures that before `testCppClient` tries to _start_ to build its own image, cppClient must be complete.

Reply:
To restate, because parentContainers= already has a reference to :cppClient, this line which I "corrected" is entirely unnecessary.